Breaking: overflow detection for SDIV - minInt / (-1)

pirapira opened the pull request:
Trying to fix #1087.

First I'm trying to catch this in the CI servers.

pirapira committed:
Add a test case for #1087

pirapira committed:
Change a name of a function in the test

Perhaps `main` has a special syntactic treatment

pirapira committed:
Oh, when I misspell `function` as `funtction`, a syntax error happens.

pirapira committed:
Add a runtime check for (minInt) / (-1)

Throw when the result is not within the expressible range of signed 256 bit integer.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1337,6 +1337,20 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		// Test for (minInt) / (-1)
+		if (c_isSigned && _operator == Token::Div)
+		{
+			m_context << Instruction::DUP1 << Instruction::NOT  // ~x
```
Not sure if this is faster, but you might want to use the new inline assembly here:

```
m_context.appendInlineAssembly(R"(
  and(
    eq(x, exp(2, 255)), // x == 2**256 (min int)
    iszero(not(y)) // y == -1
  )
)",
// names of variables already on the stack
{"y", "x"});
m_context.appendConditionalJumpTo(m_context.errorTag());
```

Oh by the way, in all other comments of the code generator, the stack grows to the right, I think yours grows to the left.

pirapira replied:
Oh, I'll change the stack growing direction.

chriseth commented:
Please don't miss my comment above.

pirapira commented:
No, no, they arrive in my mailbox.

pirapira commented:
@chriseth The inline assembly caused an exception (possibly because the compiler tried to reduce every constant number, and hit the case?)  So I reverted to the old fashioned EVM writing.

pirapira commented:
This is done on my side.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1337,6 +1337,20 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		if (c_isSigned && _operator == Token::Div)
+		{
+			m_context << Instruction::DUP1 // x
+				<< u256(255) << u256(2) << Instruction::EXP // 2 ** 255
```
Better to use `(u256(1) << 255)`, the optimizer knows how to represent that.

pirapira replied:
Does the optimizer use `EXP`, `PUSH32` directly, or a much better way?  I will do an experiment.

chriseth replied:
It tries various different ways to represent constants and compares the creation-time costs against the execution-time costs.

pirapira replied:
OK `--optimize` does the same thing as I do.  I'll change this.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1337,6 +1337,20 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+
 		if (_operator == Token::Div)
 			m_context << (c_isSigned ? Instruction::SDIV : Instruction::DIV);
 		else
```
Is there also a special case for smod?

pirapira replied:
The `SMOD` specification is [ambiguous](https://github.com/ethereum/yellowpaper/issues/190).  I created [a separate issue](https://github.com/ethereum/solidity/issues/1162) about this.

pirapira replied:
My conclusion is that `SMOD` has no overflows.

chriseth reviewed (changes requested):
Please add an entry into the changelog (not for the current version, this has to go into 0.5.0) and also add some documentation ("places where solidity throws").

Note that this PR is only one part of the general "throw on arithmetic overflow" task.

pirapira commented:
Mentioning #796 for cross-referencing.

pirapira committed:
Add Changelog entry about the new throw

that happens when the smallest signed integer is divided by one.

pirapira committed:
Let the optimizer optimize

Following a comment https://github.com/ethereum/solidity/pull/1091#discussion_r81823106

pirapira commented:
I've dealt with all comments.

axic commented:
Can we pull out the "When Solidity Throws" change into a separate PR and merge it?

pirapira commented:
@axic that's a good idea.  I start the separation.

pirapira commented:
@axic I separated #1167 and rebased this one on top of it.

pirapira committed:
Add documentation about the new throw case

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -6287,6 +6287,26 @@ BOOST_AUTO_TEST_CASE(divisiod_by_zero)
+				return a / b;
+			}
+			function test() returns (int) {
+				return div(minInt(), -1);
```
Are you using multiple methods to avoid the optimizer removing the condition? If so, can you leave a comment?

pirapira replied:
I just wanted to write types explicitly so that it is obviously an `int` divided by an `int`.

axic replied:
I understand the need of `minInt()`, otherwise it would be a literal number and hence won't become assembly.

I think `return minInt() / -1;` shouldn't be optimised out, but changing `mintInt` to `external` will definitely not let the optimiser take over. Maybe this is something we need to consider for tests.

pirapira commented:
@chriseth I've dealt with your review comments.

axic commented:
@pirapira please rebease 😉

pirapira commented:
I guess the next thing that can happen on this PR is @chriseth to approve it.

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1358,6 +1358,19 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+				<< Instruction::ISZERO;
+			// y, x, (x == 2 ** 255), (y == -1)
+			m_context << Instruction::AND;
+			m_context.appendConditionalJumpTo(m_context.errorTag());
```
Please change to invalid instruction.

pirapira replied:
I changed this.

pirapira committed:
codegen: use 0xfe opcode for minInt/(-1)

pirapira committed:
codegen: replace some instructions with an inline assembly

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1418,6 +1418,16 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		if (c_isSigned && _operator == Token::Div)
+		{
+			m_context.appendInlineAssembly(R"(
+				and(eq(x, u256(1) << 255), eq(y, -1))
```
I don't think `u256(1) << 255` works?

It has to be a supplied constant in this case or just hardcode the value.

pirapira replied:
Thanks.

pirapira committed:
Replace C++ expression in inline assembly

axic commented:
This managed to trigger an error:

```
Testing soljson via the fuzzer...
Invalid error: "Internal compiler error (/home/travis/build/ethereum/solidity/libsolidity/codegen/CompilerContext.cpp:295): Failed to assemble inline assembly block.
"
scripts/../test/cmdlineTests.sh: line 60: 10730 Aborted                 (core dumped) 
```

pirapira commented:
So it might be the case the end-to-end test is not working.

chriseth commented:
There is an eq opcode with three arguments. I don't think we have to fix this now, there is a lot to be done with all the other operations here.

pirapira commented:
I somehow prioritized this.

pirapira commented:
Tests pass finally.   I'll keep rebasing.

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1450,6 +1450,16 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		if (c_isSigned && _operator == Token::Div)
+		{
+			m_context.appendInlineAssembly(R"({
+				jumpi(invalidJumpLabel, and(eq(y, exp(2,255)), iszero(add(x, 1))))
```
The `invalidJumpLabel` is gone. This should use something like:
```
jumpi(safe, iszero(...))
invalid
safe:
```

Thought not sure if the labels must be unique?

pirapira replied:
I'll create something like
```
invalidLabel:
invalid
```
once somewhere.

It seems I can do something within 3 weeks.

chriseth commented:
Overflow-detection at runtime is extremely complicated because of the various types we have. I propose to do only compile-time overflow checking and close this pull request.

pirapira commented:
@chriseth do you have less number of types at compilation time?

chriseth commented:
No, but it is not as complicated and the effects of getting something wrong are not as bad.

pirapira commented:
@chriseth really?  I think runtime checks are vastly less complicated than compilation time checks.

pirapira commented:
@chriseth brought up a convincing argument that all different cases of the runtime overflow checks will not make it until the new intermediate language lands. So I'm closing this.
